In [57]:
import pandas as pd
import numpy as np

In [58]:
df=pd.read_csv("car_fuel_efficiency.csv")
df.head(10)

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   
5                  190            3.0         NaN     2484.883986   
6                  240            7.0       127.0     3006.542287   
7                  150            4.0       239.0     3638.657780   
8                  250            1.0       174.0     2714.219310   
9                  150            4.0       123.0     3509.036569   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   
5          14.7        2008  Europe  Gasoline    All-wheel drive       -1.0   
6          22.2        2012     USA  Gasoline  Front-wheel drive        1.0   
7          17.3        2020     USA    Diesel    All-wheel drive        1.0   
8          10.3        2016    Asia    Diesel  Front-wheel drive       -1.0   
9          10.2        2005     USA  Gasoline  Front-wheel drive       -1.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369  
5            17.271818  
6            13.210412  
7            12.848884  
8            16.823554  
9            12.298355

reparing the dataset
Preparation:

Fill missing values with zeros.
Do train/validation/test split with 60%/20%/20% distribution.
Use the train_test_split function and set the random_state parameter to 1.
Use DictVectorizer(sparse=True) to turn the dataframes into matrices.

In [59]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [60]:
df.fillna(0)


engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            0.0        78.0     3079.038997   
3                     220            4.0         0.0     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            0.0       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003  Europe  Gasoline    All-wheel drive   
1             17.8        2007     USA  Gasoline  Front-wheel drive   
2             15.1        2018  Europe  Gasoline  Front-wheel drive   
3             20.2        2009     USA    Diesel    All-wheel drive   
4             14.4        2009  Europe  Gasoline    All-wheel drive   
...            ...         ...     ...       ...                ...   
9699          17.3        2013  Europe    Diesel  Front-wheel drive   
9700          15.0        2004     USA  Gasoline    All-wheel drive   
9701          15.1        2008     USA    Diesel    All-wheel drive   
9702          19.4        2011     USA    Diesel  Front-wheel drive   
9703          14.7        2005    Asia    Diesel    All-wheel drive   

      num_doors  fuel_efficiency_mpg  
0           0.0            13.231729  
1           0.0            13.688217  
2           0.0            14.246341  
3           2.0            16.912736  
4           2.0            12.488369  
...         ...                  ...  
9699        0.0            15.101802  
9700        0.0            17.962326  
9701       -1.0            17.186587  
9702        1.0            15.331551  
9703       -1.0            14.884467  

[9704 rows x 11 columns]

In [61]:
df.fillna(0,inplace=True)
df.isnull().sum()

engine_displacement    0
num_cylinders          0
horsepower             0
vehicle_weight         0
acceleration           0
model_year             0
origin                 0
fuel_type              0
drivetrain             0
num_doors              0
fuel_efficiency_mpg    0
dtype: int64

In [62]:
from sklearn.model_selection import train_test_split
df_full_train,df_test=train_test_split(df,test_size=0.2,random_state=1)
df_train,df_val=train_test_split(df_full_train,test_size=0.25,random_state=1)


In [63]:
df_train = df_train.reset_index(drop=True)
df_val = df_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

In [64]:
y_train=df_train['fuel_efficiency_mpg'].values
y_test=df_test['fuel_efficiency_mpg'].values
y_val=df_val['fuel_efficiency_mpg'].values

In [65]:
del df_train['fuel_efficiency_mpg']
del df_test['fuel_efficiency_mpg']
del df_val['fuel_efficiency_mpg']

In [66]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.feature_extraction import DictVectorizer
from sklearn.metrics import roc_auc_score

In [67]:
train_dicts = df_train.fillna(0).to_dict(orient='records')
train_dicts[:5]

[{'engine_displacement': 120,
  'num_cylinders': 5.0,
  'horsepower': 169.0,
  'vehicle_weight': 2966.67950463456,
  'acceleration': 13.9,
  'model_year': 2005,
  'origin': 'USA',
  'fuel_type': 'Gasoline',
  'drivetrain': 'Front-wheel drive',
  'num_doors': -1.0},
 {'engine_displacement': 200,
  'num_cylinders': 3.0,
  'horsepower': 143.0,
  'vehicle_weight': 2950.822121334588,
  'acceleration': 17.1,
  'model_year': 2013,
  'origin': 'Asia',
  'fuel_type': 'Diesel',
  'drivetrain': 'Front-wheel drive',
  'num_doors': -1.0},
 {'engine_displacement': 180,
  'num_cylinders': 6.0,
  'horsepower': 180.0,
  'vehicle_weight': 3078.221669239548,
  'acceleration': 17.4,
  'model_year': 2007,
  'origin': 'USA',
  'fuel_type': 'Gasoline',
  'drivetrain': 'All-wheel drive',
  'num_doors': 0.0},
 {'engine_displacement': 280,
  'num_cylinders': 5.0,
  'horsepower': 174.0,
  'vehicle_weight': 2797.9917925588984,
  'acceleration': 0.0,
  'model_year': 2016,
  'origin': 'USA',
  'fuel_type': 'Diesel'

In [68]:
dv = DictVectorizer(sparse=True)
X_train = dv.fit_transform(train_dicts)
X_train

<5822x14 sparse matrix of type '<class 'numpy.float64'>'
	with 58220 stored elements in Compressed Sparse Row format>

Question 1
Let's train a decision tree regressor to predict the fuel_efficiency_mpg variable.

Train a model with max_depth=1.
Which feature is used for splitting the data?
the answer is :vehicle_weight

In [69]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error

Question 2
Train a random forest regressor with these parameters:

n_estimators=10
random_state=1
n_jobs=-1 (optional - to make training faster)
What's the RMSE of this model on the validation data?

In [70]:
dt = DecisionTreeRegressor(max_depth=1)
dt.fit(X_train, y_train)
 
val_dicts = df_val.fillna(0).to_dict(orient='records')
X_val = dv.transform(val_dicts)
y_pred=dt.predict(X_val)

In [71]:

from sklearn.metrics import root_mean_squared_error

rmse = root_mean_squared_error(y_val, y_pred)
print("RMSE:", rmse)


RMSE: 1.6104639028827594


In [72]:
feature_names = dv.get_feature_names_out()
split_feature_index = dt.tree_.feature[0]
split_feature_index
print("Feature utilisée :", feature_names[split_feature_index])


Feature utilisée : vehicle_weight


In [73]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(
    n_estimators=10,
    random_state=1,
    n_jobs=-1   # utilise tous les coeurs CPU (optionnel)
)

rf.fit(X_train, y_train)

y_pred = rf.predict(X_val)
y_pred


array([18.62692845, 15.23747064, 18.20763942, ..., 14.81525146,
       13.52753444, 16.10909938])

In [74]:
from sklearn.metrics import root_mean_squared_error

rmse = root_mean_squared_error(y_val, y_pred)
print("RMSE:", rmse)



RMSE: 0.4595777223092726


The RMSE is 0.45

Question 3
Now let's experiment with the n_estimators parameter

Try different values of this parameter from 10 to 200 with step 10.
Set random_state to 1.
Evaluate the model on the validation dataset.
After which value of n_estimators does RMSE stop improving? Consider 3 decimal places for calculating the answer.

10
25
80
200

The RMSE stop mproving after 80

In [75]:
scores=[]
for n in range(10,201,10):
    rf = RandomForestRegressor(
    n_estimators=n,
    random_state=1,
    n_jobs=-1 
    )
    rf.fit(X_train,y_train)
    y_pred=rf.predict(X_val)
    rmse = root_mean_squared_error(y_val, y_pred)
    scores.append((n,rmse))



Question 4
Let's select the best max_depth:

Try different values of max_depth: [10, 15, 20, 25]
For each of these values,
try different values of n_estimators from 10 till 200 (with step 10)
calculate the mean RMSE
Fix the random seed: random_state=1
What's the best max_depth, using the mean RMSE?

10
15
20
25

In [76]:
scores=[]
for s in [10, 15, 20, 25]:
    for n in range(10,201,10):
        rf = RandomForestRegressor(
    n_estimators=n,
    random_state=1,
    max_depth=s,
    n_jobs=-1 
    )
    rf.fit(X_train,y_train)
    y_pred=rf.predict(X_val)
    rmse = root_mean_squared_error(y_val, y_pred)
    scores.append((s,n,rmse))


In [77]:
scores

[(10, 200, 0.43984510625501455),
 (15, 200, 0.44234558481369435),
 (20, 200, 0.44290498719773974),
 (25, 200, 0.44247941055635037)]

The best max_deep is 10 

Question 5

In [78]:
rf = RandomForestRegressor(
    n_estimators=10,
    random_state=1,
    max_depth=20,
    n_jobs=-1 
    )
rf.fit(X_train,y_train)
y_pred=rf.predict(X_val)
rmse = root_mean_squared_error(y_val, y_pred)


In [79]:
importances = rf.feature_importances_
importances



array([1.14797006e-02, 3.57108549e-04, 3.45384113e-04, 3.27279191e-03,
       3.25424323e-04, 3.60383601e-04, 1.59978977e-02, 3.21230009e-03,
       2.34334695e-03, 1.63498954e-03, 4.62246496e-04, 5.18739639e-04,
       5.39721689e-04, 9.59149965e-01])

In [80]:
import pandas as pd

importances = rf.feature_importances_
feature_names = dv.get_feature_names_out()

feat_imp = pd.DataFrame({
    'feature': feature_names,
    'importance': importances
}).sort_values(by='importance', ascending=False)

print(feat_imp)


                         feature  importance
13                vehicle_weight    0.959150
6                     horsepower    0.015998
0                   acceleration    0.011480
3            engine_displacement    0.003273
7                     model_year    0.003212
8                  num_cylinders    0.002343
9                      num_doors    0.001635
12                    origin=USA    0.000540
11                 origin=Europe    0.000519
10                   origin=Asia    0.000462
5             fuel_type=Gasoline    0.000360
1     drivetrain=All-wheel drive    0.000357
2   drivetrain=Front-wheel drive    0.000345
4               fuel_type=Diesel    0.000325


The best feature_importance is: vehicle_weight    0.959150

uestion 6
Now let's train an XGBoost model! For this question, we'll tune the eta parameter:

Install XGBoost
Create DMatrix for train and validation
Create a watchlist
Train a model with these parameters for 100 rounds:
xgb_params = {

In [82]:
!pip install xgboost

In [83]:
import xgboost as xgb


In [84]:
dtrain = xgb.DMatrix(X_train, label=y_train)
dval = xgb.DMatrix(X_val, label=y_val)

# 5. Create a watchlist
watchlist = [(dtrain, 'train'), (dval, 'eval')]

# 6. Define parameters
xgb_params = {
    'eta': 0.3, 
    'max_depth': 6,
    'min_child_weight': 1,
    'objective': 'reg:squarederror',
    'nthread': 8,
    'seed': 1,
    'verbosity': 1,
}

# 7. Train the model for 100 rounds
model = xgb.train(
    params=xgb_params,
    dtrain=dtrain,
    num_boost_round=100,
    evals=watchlist
)

# 8. Make predictions on validation set
y_pred = model.predict(dval)
print("\nSample predictions:", y_pred[:5])

[0]	train-rmse:1.81393	eval-rmse:1.85444
[1]	train-rmse:1.31919	eval-rmse:1.35353
[2]	train-rmse:0.98120	eval-rmse:1.01316
[3]	train-rmse:0.75443	eval-rmse:0.78667
[4]	train-rmse:0.60680	eval-rmse:0.64318
[5]	train-rmse:0.51381	eval-rmse:0.55664
[6]	train-rmse:0.45470	eval-rmse:0.50321
[7]	train-rmse:0.41881	eval-rmse:0.47254
[8]	train-rmse:0.39534	eval-rmse:0.45509
[9]	train-rmse:0.38038	eval-rmse:0.44564
[10]	train-rmse:0.37115	eval-rmse:0.43896
[11]	train-rmse:0.36361	eval-rmse:0.43594
[12]	train-rmse:0.35850	eval-rmse:0.43558
[13]	train-rmse:0.35365	eval-rmse:0.43394
[14]	train-rmse:0.35025	eval-rmse:0.43349
[15]	train-rmse:0.34666	eval-rmse:0.43362
[16]	train-rmse:0.34459	eval-rmse:0.43378
[17]	train-rmse:0.34128	eval-rmse:0.43405
[18]	train-rmse:0.33822	eval-rmse:0.43391
[19]	train-rmse:0.33709	eval-rmse:0.43374
[20]	train-rmse:0.33553	eval-rmse:0.43376
[21]	train-rmse:0.33243	eval-rmse:0.43453
[22]	train-rmse:0.33031	eval-rmse:0.43510
[23]	train-rmse:0.32815	eval-rmse:0.43601
[2

In [85]:
dtrain = xgb.DMatrix(X_train, label=y_train)
dval = xgb.DMatrix(X_val, label=y_val)

# 5. Create a watchlist
watchlist = [(dtrain, 'train'), (dval, 'eval')]

# 6. Define parameters
xgb_params = {
    'eta': 0.1, 
    'max_depth': 6,
    'min_child_weight': 1,
    'objective': 'reg:squarederror',
    'nthread': 8,
    'seed': 1,
    'verbosity': 1,
}

# 7. Train the model for 100 rounds
model = xgb.train(
    params=xgb_params,
    dtrain=dtrain,
    num_boost_round=100,
    evals=watchlist
)

# 8. Make predictions on validation set
y_pred = model.predict(dval)
print("\nSample predictions:", y_pred[:5])

[0]	train-rmse:2.28944	eval-rmse:2.34561


[1]	train-rmse:2.07396	eval-rmse:2.12434
[2]	train-rmse:1.88066	eval-rmse:1.92597
[3]	train-rmse:1.70730	eval-rmse:1.74987
[4]	train-rmse:1.55163	eval-rmse:1.59059
[5]	train-rmse:1.41247	eval-rmse:1.44988
[6]	train-rmse:1.28796	eval-rmse:1.32329
[7]	train-rmse:1.17660	eval-rmse:1.20930
[8]	train-rmse:1.07736	eval-rmse:1.10830
[9]	train-rmse:0.98883	eval-rmse:1.02009
[10]	train-rmse:0.91008	eval-rmse:0.94062
[11]	train-rmse:0.84030	eval-rmse:0.87100
[12]	train-rmse:0.77874	eval-rmse:0.80916
[13]	train-rmse:0.72417	eval-rmse:0.75465
[14]	train-rmse:0.67626	eval-rmse:0.70780
[15]	train-rmse:0.63402	eval-rmse:0.66672
[16]	train-rmse:0.59690	eval-rmse:0.63062
[17]	train-rmse:0.56447	eval-rmse:0.60016
[18]	train-rmse:0.53619	eval-rmse:0.57383
[19]	train-rmse:0.51138	eval-rmse:0.55044
[20]	train-rmse:0.48983	eval-rmse:0.53064
[21]	train-rmse:0.47135	eval-rmse:0.51451
[22]	train-rmse:0.45501	eval-rmse:0.49998
[23]	train-rmse:0.44120	eval-rmse:0.48790
[24]	train-rmse:0.42929	eval-rmse:0.47773
[

Which eta leads to the best RMSE score on the validation dataset? the answer is:0.1
